src/tardev-snapshotter/src/snapshotter.rs

```diff
@@ -887,63 +887,110 @@ impl TarDevSnapshotter {
 
             let upstream_name = base_name.with_extension(layer_type);
 
-            info!("Fetching {} layer image to {:?}", layer_type, upstream_name);
-            self.get_layer_image(&upstream_name, digest_str).await?;
-
-            // Process the layer
-            let generated_root_hash = tokio::task::spawn_blocking(move || -> Result<_> {
-                if layer_type == TAR_EXTENSION {
-                    info!("Renaming {:?} to {:?}", &upstream_name, &base_name);
-                    std::fs::rename(&upstream_name, &base_name)?;
-                }
-                let mut file = OpenOptions::new()
-                    .read(true)
-                    .write(true)
-                    .create(true)
-                    .truncate(layer_type == TAR_GZ_EXTENSION)
-                    .open(&base_name)?;
-                if layer_type == TAR_GZ_EXTENSION {
-                    info!("Decompressing {:?} to {:?}", &upstream_name, &base_name);
-                    let compressed = fs::File::open(&upstream_name).map_err(|e| {
-                        let file_error = format!(
-                            "Failed to open file {:?} for decompression: {:?}",
-                            &upstream_name, e
-                        );
-                        error!("{}", file_error);
-                        anyhow::anyhow!(file_error)
-                    })?;
-
-                    let mut gz_decoder = flate2::read::GzDecoder::new(compressed);
-
-                    if let Err(e) = std::io::copy(&mut gz_decoder, &mut file) {
-                        let copy_error = format!("failed to copy payload from gz decoder {:?}", e);
-                        error!("{}", copy_error);
-                        return Err(anyhow::anyhow!(copy_error));
+            const MAX_RETRIES: usize = 3;
+            let mut retries = 0;
+
+            while retries < MAX_RETRIES {
```

Comment:
Can you add a comment into the code as to why we believe we need this while loop for retrying? I am worried we still dont understand the root cause of the issue, and so would like to leave the context for the future readers.

Reply:
Yes. Comments added. Retry will help us debug more if we hit this error again.